Sam's asm and ammx optims

arczi84 opened the pull request.

Samuel-DEVULDER committed:
added amiga files

Samuel-DEVULDER committed:
fixed Configure.readme script

Samuel-DEVULDER committed:
1st successfull compilation

Samuel-DEVULDER committed:
fixed palette + intro-movie

Samuel-DEVULDER committed:
moved fps to the right in case screen truncate left column

Samuel-DEVULDER committed:
added asm version of render

Samuel-DEVULDER committed:
SAGA started to work!

Samuel-DEVULDER committed:
AMMX speedup

Samuel-DEVULDER committed:
fixed random crashes due to SDL_library data being trashed

Samuel-DEVULDER committed:
oops a file was missing

Samuel-DEVULDER committed:
Speedup of render2 with Mask

Samuel-DEVULDER committed:
now uses E4/E5 in place of D4/D5 by usage of BANK prefix

Samuel-DEVULDER committed:
made an AMMX version of memcpy

Samuel-DEVULDER committed:
fixed compilation issue

Samuel-DEVULDER committed:
now decodes smk bitstream with a faster algorithm ==> video at intros now plays fine on the vamp

Samuel-DEVULDER committed:
code cleanup

Samuel-DEVULDER committed:
cleanup

Samuel-DEVULDER committed:
fixed random crash in intro-video

Samuel-DEVULDER committed:
speedup smk with a bit of 68k asm.

Samuel-DEVULDER committed:
at least a working SAGA version that doesn't modify internal SDL data. Movie intro now works fine, at last!

Samuel-DEVULDER committed:
added AMMX replacement for memcpy/memset.

Samuel-DEVULDER committed:
Better handling of .gitignore to include amiga asm files

Samuel-DEVULDER committed:
more asm optim + support for profiling in smk

Samuel-DEVULDER committed:
improved branch prediction in RenderTile_RT_TRANSPARENT

Samuel-DEVULDER committed:
added asm replacement for memcmp().

Samuel-DEVULDER committed:
speedup

Samuel-DEVULDER committed:
fixed typo

Samuel-DEVULDER committed:
fixed typo

Samuel-DEVULDER committed:
fixed typo

Samuel-DEVULDER committed:
now uses dbmalloc()/dbfree() in smk resulting in faster video-intro

Samuel-DEVULDER committed:
Merge branch 'malloc-replacement'

Samuel-DEVULDER committed:
now uses a 16bits buffer ==> works even faster

Samuel-DEVULDER committed:
get rid of SDL clipping when rendering video

Samuel-DEVULDER committed:
typo

Samuel-DEVULDER committed:
slight speedup in smk

Samuel-DEVULDER committed:
Ctrl-C now should work even when menus are active

Samuel-DEVULDER committed:
improved ctrl-c support when in character selection menu

Samuel-DEVULDER committed:
make StormLib use dlmalloc

Samuel-DEVULDER committed:
Now automatically check that stack is at least 128kb (experimental minimum being around 64kb)

Samuel-DEVULDER committed:
speedup: now usess regparm(2) for smk bitstreams fcn

Samuel-DEVULDER committed:
speedup WastBit (most used fcn in there) with a bit of inline asm

Samuel-DEVULDER committed:
asm + ammx speedup

MBeijer commented:
@ajenbo

MBeijer commented:
@Samuel-DEVULDER ping

Samuel-DEVULDER commented:
pong! 

I can see a red message stating "merging is blocked since review is required" . But at least 2 approving reviews are requested.

What should I do at this point ?

Samuel-DEVULDER reviewed (dismissed):
These look like my changes :)

AJenbo commented:
A PR of this size an nature should really be accompanied by some description of what it does and what it archives (does it get the Amiga 500 from 34fps to 39fps?, does it make specific parts compatible with more hardware? Why was it done in this way?). Normally I get grumpy when a PR is +1000 lines, this one is almost 10x that size and as such close to impossible to review and it's highly specialized nature makes it hard to maintain (can we expect some one to maintain this, we would like to establish a relation ship before accepting all this).

Since a lot of this is highly system specific I'm not sure how well suited it is for upstreaming. Especially when we still have some major rewrites planned for the render. I suggest you relegate those changes to a separate branch.

Maybe try upstreaming some of the changes to libsmacker https://github.com/JonnyH/libsmacker and stormlib https://github.com/ladislav-zezula/StormLib so that DevilutionX doesn't have to carry the patches.

With all that said this is very interesting and some commits looks very attractive for upstreaming directly (8bit color fixes etc). If they could be split in to individual PRs that would make things much easier to go over.

Samuel-DEVULDER commented:
I understand. So let me explain...

These changes regoups ~18days of optimizing (fps raised from ~20 to ~36 on my vamp and video playback is near to perfect).

They can be read as follows:

1. Some random/crash appears on the amiga version because of a race condition in which a semaphore is used after being released resulting in internal SDL_library data being trashed.

	Solution: add a flag to indicated that the mutex is valid or not.
	
	Impact: changes in 3rdParty/Storm/Source/storm.h https://github.com/AmigaPorts/devilutionX/pull/46/commits/97d0b67df890bb16a40ad929837c78dbaf9afab8

2. Added 68k asm version of render.cpp code: 

	Impact: small C change + new files support.c + render68k.asm + vampire includes in a platform-specific folder + modification of CMake files to allow compilation of asm) https://github.com/AmigaPorts/devilutionX/pull/46/commits/f05cbc3e68a7feef68571d218d77c63e5634af12

3. Added SAGA (68080 specific) support to allow faster RenderPresent().

	Impact: small changes in render.cpp (macro to replace SDL_Flip with a replacement) + modifications of amiga-specific support.c https://github.com/AmigaPorts/devilutionX/pull/46/commits/97d0b67df890bb16a40ad929837c78dbaf9afab8#diff-793273cdae5d9b9c4b0d58de22e1341e
	
4. After profiling it appears that

	4. a) memcpy/memset/memcpy are heavily used by the program and they are not 68080 optimized. 
	   
		Solution: replace these by optimized "asm-wrappers" (see ld --wrap=symbol) on link time.

		Impact: new asm file containing the memory related functions (memopt.asm). No C/CPP files changed because of the used of gcc's wrap feature. https://github.com/AmigaPorts/devilutionX/pull/46/commits/4d9005c51d5aa77ad5522060a5075d7b9fd63cfa
			
	4. b) smk's bitstream reading functions are used very often and not well optimized by GCC. 
	   
		Solution: reswork bitstream internal data structure to make it less costly (reduced array access using internal 16bit buffer + C and 68k-asm version of the routines).
	   
		Impact: smk's internal smk_bitstream.c/h (api still compatible with  original code). https://github.com/AmigaPorts/devilutionX/pull/46/commits/81c0c9760fbf9ea51c7b50e0069751eadd076874
	   
	4. c) smk's huffman tree routines are badly optimized by GCC on 68k.
	
		Solution: add inline asm version of them
		
		impact: smk_hufftree.c https://github.com/AmigaPorts/devilutionX/pull/46/commits/89acfd3f1b80eb89c88e4bcbbf6512e40ba8ab60#diff-e5102224a3f26efb4178acce2361071e
	
	4. d) free() uses lots of cpu time (>14%) during video playback. 
	
		Solution: Use Doug Lea's memory management routine for the playback. With that code, dlfree() now uses less than 1.4% of cpu-time (10x faster that libnix memory management system).

		Impact: added Doug Lea's malloc.c in amiga-specific folder + modified smk_malloc.h to use it. https://github.com/AmigaPorts/devilutionX/pull/46/commits/bcac2487d8d91dbf45ee1c52b10a1276e3208bad
	
Well, as far as I can tell, this is all of the story behind these changes.

Samuel-DEVULDER commented:
oops it seem I have trouble with indexing.. iv.a) v.b) etc. should have been 4.a) 4.b) etc.
(I'm new to using these)

I also forget to tell these more recent changes

1. Make storm_lib also use Doug's Lea malloc system to allow kind of central unification of the memory used by the game (ctrl-E on console display memory usage statistics by the program) https://github.com/AmigaPorts/devilutionX/pull/46/commits/cc465201d012c4a00d871ea2c58e15e98636deee

1. Make program gracefully exit on ctrl-c https://github.com/AmigaPorts/devilutionX/pull/46/commits/1ef3c13bea275c78907965bafe38d5ee62c8a1ab https://github.com/AmigaPorts/devilutionX/pull/46/commits/3faf0207123016814903fd2eb50b46317ffea15b

1. Added code to fix crashes caused by stack being too small https://github.com/AmigaPorts/devilutionX/pull/46/commits/ff3dc3fcb6d88524879564c74f18e43841a00d86

1. Various recent optimization by using regparm(2) api https://github.com/AmigaPorts/devilutionX/pull/46/commits/8282761f7111a75519c27c18dc16fa932577a76a

1. Optimize  PKWare WasteBit() function to speedup loading times. https://github.com/AmigaPorts/devilutionX/pull/46/commits/e3ba799032b5adb7cbfed2f9d819d79a426acfe2

**notice**: All of my changes are amiga-specific and should not affect other targets (#ifdef protection used everywere. )

AJenbo commented on Source/init.cpp:
```diff
@@ -47,7 +47,7 @@ void init_create_window()
 		app_fatal("Unable to create main window");
 	dx_init(NULL);
-	atexit(dx_cleanup);
+	//atexit(dx_cleanup);
```
This will defiantly affect other targets :)

Samuel-DEVULDER replied:
Right! I added it upon suggestion From Arczi84 i think because it seem to fix issues on the amiga. But actually the issue was in the freed semaphore being reused. Core review proves to be usefull once again because this changes should'nt be there. I'll correct my branch. Thank you for pointing this out.

BTW, are you dealing with the "AmigaDev" repo or the "Diasurgical" one above for this PR? (I'd tought Arczi juste send a PR to "AmigaDev" so that he can sync with my changes).

AJenbo replied:
My comments are in regards to merging thing in to Diasurgical, we have a goal of being fully BE compatible, so basic Amiga support would be a good way for us to achieve this. There is currently a PR for merging AmigaDev:master in to Diasurgical and as it it's good except for 3 bugs (save game support, crash and bad colors), so I have a bit of interest in the AmigaDev staying in a state where it can be merged in to Diasurgical :)

https://github.com/diasurgical/devilutionX/pull/397

AJenbo commented on Source/scrollrt.cpp:
```diff
@@ -1061,7 +1062,7 @@ static void DrawFPS()
 		}
 		wsprintf(String, "%d FPS", framerate);
-		PrintGameStr(8, 65, String, COL_RED);
+		PrintGameStr(8+8, 65, String, COL_RED);
```
Could you elaborate what is truncating this part of the screen?

Samuel-DEVULDER replied:
On my tv/monitor (LG), the left and right 8 pixels of any 640x480 screens is not displayed. This was a tmp fix for my setup. I shouldn't have make this leaked out but I don't know if it is possible to mark some specific changed lines as "invisible" for git.

AJenbo replied:
Ok that's what I expected :) I normally commit lines/block individually which helps catch stuff like this. But I also realize that you didn't make this PR so your might not have felt that it was totally ready, and you didn't really have a chance to write a summery either, but you have been very helpful in commenting on the changes here in the review so thank you for that :)

Samuel-DEVULDER committed:
restored removed atexit

AJenbo commented on SourceS/sdl_compat.h:
```diff
@@ -94,6 +80,12 @@ inline int SDLC_SetSurfaceAndPaletteColors(SDL_Surface *surface, SDL_Palette *pa
+	#if SDL1_VIDEO_MODE_BPP == 8
+		// In SDL1, if the Video Surface is 8bit, you need to set this as well to match,
+		// otherwise colors will be wrong!
+		SDL_SetColors(SDL_GetVideoSurface(), colors, firstcolor, ncolors);
```
Could this be solved using `SDL_SetPalette(surface, SDL_LOGPAL | SDL_PHYSPAL, ...` instead?
https://www.libsdl.org/release/SDL-1.2.15/docs/html/sdlsetcolors.html

Samuel-DEVULDER replied:
Don't really know. This is one of the first change I made upon Arczi's advice in the beginning to get the colors right. From my inspecting SDL sources, it seem that the surface palette is'nt automatically copied onto the VideoSurface (missing a version counter in palette maybe?), and then SDL does approximates the surface colors.

Let me test with the SDL_PHYSPAL as well...

Nope: it doesn't work with SDL_SetColors commented-out:
![image](https://user-images.githubusercontent.com/33961444/71269156-edaa3e80-234e-11ea-8cf1-9f44023cd1bb.png)
![image](https://user-images.githubusercontent.com/33961444/71269182-fdc21e00-234e-11ea-8149-f56bbe511bcb.png)

AJenbo replied:
Ok, in that case we can upstream that commit directly. Can you also take a look at SDLC_SetSurfaceColors since it has also been made to use SDL_PHYSPAL.

AJenbo replied:
Thanks for going over this, the fix has now been cherry-picked upstream:
https://github.com/diasurgical/devilutionX/commit/319bdeddd242bb79c5145fd81ee68d584d4b220e

AJenbo commented on SourceX/storm/storm.cpp:
```diff
@@ -703,7 +712,9 @@ BOOL SVidPlayContinue(void)
 #endif
-		ScaleOutputRect(&pal_surface_offset);
+#ifndef __AMIGA__
+		ScaleOutputRect(&pal_surface_offset); // somehow this lead to pb with amiga's SDL
```
pb?

Samuel-DEVULDER replied:
hmm.. yeah I got strange result with that that call at one point in time. But with changes between lines 677-684 this is no more needed apparently.

AJenbo replied:
Well I'm just not sure what "pb" stands for in this case :)

Samuel-DEVULDER replied:
pb = problem
but I think we can remove the change because when I do so, the exe seem to work fine under uae.

Samuel-DEVULDER committed:
made change amiga-specific

Samuel-DEVULDER committed:
made change amiga-specific

AJenbo commented:
Regarding the Amiga specific optimizations. The memcpy/memset sounds nice an clean, and I'm not to much opposed to the work done on libsmacker either. Of-cause it's more idea if any of the changes could be made in auxiliary projects (like SDL).

On review I agree that you have done a fairly nice job of keeping things contained so that there should be minimal impact. But things might still break when we rewrite the render, not for other platforms but for Amiga, so for that reason I would still request that that part be left out. It can either be reintroduced once we split DevilutionX in a 1.0.x and 1.x.y branch, or once the render has been rewritten. Also if the C code is already hitting 20fps the game would be equal to the original PC release as that was the cap, and it may improve further once we rewrite the render to use SDL routines instead of all the custom pixel pushing currently employed.

@Samuel-DEVULDER Nice work on this.

Samuel-DEVULDER commented:
Actually it is the render rework which has the most impact on the gameplay and that Arczi was expected. 

Notice that the 20fps I was refering to was the speed on the vamp which runs routhly twice the spead of the 68060@50mhz. This mean that on non-vampirized amiga the game was running very very slow giving bad user-experience. The asm changes I made in render speed up all 68k amigas but I haven't been able to measure the impact on a non-vampirized amiga. On the vamp it nearly doubles the speed of the game "in town" (a bit less in the dungeon because of the heavy use of light-tables.)

Also notice that the change in render.cpp is rather local. All these functions are leaves, meaning that render.cpp can be rewritten at will without further consequences (apart from big fps drop because human-ASM does a better job at scheduling instructions on this part than GCC.) If render is changed, it'll be normal to rewrite the asm part if needed. 
![image](https://user-images.githubusercontent.com/33961444/71271154-89d64480-2353-11ea-8ca9-183e35c423da.png)

AJenbo commented:
> Actually it is the render rework which has the most impact on the gameplay and that Arczi was expected.

I don't think we are quite on the same page here. By "render" I'm referring to render.cpp, scrollrt.cpp, engine.cpp, and a few loose functions. The previous clean up of the render was to deduplicate it and make it easier to rework, so yeah that had an expected drop in performance. But the next major rewrite is expected to have an increase in performance. And will it likely drop most of what is there currently, so it would be counter-productive to merge new code that increases it's size since it's not likely to stay with us for more than one cycle.

So right now our (DevilutionX) goal is to get basic Amiga support merged so that we can make sure that it works and that we do not break it. Once the rewrite is done, and we are happy with things in general, it makes a lot more sense to look at what optimizations can be made for specific systems.

AJenbo reviewed (changes requested):
To optimizations here are nicely segregated from the rest of the code so even if it's merged into AmigaPorts:master I don't think it would be too much of a problem to still do a selective merge into Diasurgical. The only thing that I would reguest be changed in that case is that the commits do not reformat lines that aren't actually changed (mostly the case in libsmacker).

Samuel-DEVULDER commented:
Do you have a idea of how long a cycle last? Because I'l vezry interre"sted at studying the next algorithm if it is faster.

Samuel-DEVULDER committed:
Merge branch 'master' of https://github.com/AmigaPorts/devilutionX

AJenbo commented:
2 months seams to be the norm

AJenbo commented:
>     1. Solution: add a flag to indicated that the mutex is valid or not.

Do you feel this might be worth upstreaming or is is very Amiga specific?

Samuel-DEVULDER committed:
more amiga / mc68k improvements

MBeijer commented:
I think it would be good to add this here, just so we have some logs of what is happening here:
> [23:08:32] ( __sam__) Hello
> [23:08:43] ( __sam__) I have discovered an horrible thing !!!
> [23:10:35] ( __sam__) the --allow-multiple-definition passed to the linker does something strange.. or maybe it is the extern "C" { vars declarion here }... Well ac68080_ammx was defined twice. One is for the C/C++ part of the code, the other one is for the ASM part of the code. So I have a bad and good news related to that:
> [23:12:11] ( __sam__) 1) the asm tested the "asm" version of that var, which remains "0" since it is C that changes its value. Therefore... none of the AMMX rendering / memcpy code was really executed for 2 weeks or so.. That was the bad news
> [23:13:18] ( __sam__) 2) the good news is that.. welll... the 36-40fps I have was only with plain 68k code. This mean we can go higher
> [23:15:44] ( __sam__) and between both : 1.5) it seem some of my ammx code now crashes when I unify those two variables. I'll investigate on this tomorrow (I hope).. but it was a real surprise when I realized that  even when the ASM was refering to the extern var (supposed to be the one in the C/CPP plane), the linker actually creates an anoinymous var for it without warning. This is very very odd, and I think it is related to the --allow-multiple-definition.

> [14:17:07] ( Marlon_) oopsie :D
> [14:17:19] ( Marlon_) I never use --allow-multiple-definition
> [14:17:46] ( Marlon_) it's an ugly hack when libs define the same function twice, and it will result in bad stuff like this.
> [14:18:31] ( Marlon_) it's better to fix the problem than go around it, and going around it is exactly what --allow-multiple-definition does in the linker.

Samuel-DEVULDER commented:
I have fixed the double symbol issue locally.

But now it seem that AMMX code is broken. I'm tracking all AMMX issues. Some are related to vasm not calculating the good offset when using `beq.l __real_symbol` with `-ld,wrap=symbol` feature or AMMX instructions. This is a tedious task, but I hope I'll be able eventually to have a proper branch localy that works fine with ammx.

Samuel-DEVULDER committed:
bug https://github.com/Samuel-DEVULDER/devilutionX/issues/1 fixed

Samuel-DEVULDER commented:
okay issue https://github.com/Samuel-DEVULDER/devilutionX/issues/1 is now fixed. I have a diablo running at 43fps "in town" (start point) and > 30fps in dungeon at the moment.

Samuel-DEVULDER commented:
> > ```
> > 1. Solution: add a flag to indicated that the mutex is valid or not.
> > ```
> 
> Do you feel this might be worth upstreaming or is is very Amiga specific?

Oh.. I see my answer by mail isn't appearing here. So here it is now:

> It is hard to tell because I haven't tested other platform. Somehow amiga api is weakly protected and accessing a freed resource lead to a random crash later whereas on other platform it is likely to just fail in a controlled manner, possily even silently (no check on errno here).
> 
> So I tend to think this is more important on amigaos than on other platform.
> 
> Do you think it is possible to ask "other platform" to check using this flag + printf() if they do observe use of the critical section after it has been destroyed (eg when flag says it is invalid) ?
>

Samuel-DEVULDER committed:
better handling of 'copy_pane_mask'

Samuel-DEVULDER committed:
bank corrections

Samuel-DEVULDER committed:
disable VAAPIRE_2 by default (slower code)

Samuel-DEVULDER committed:
Merge branch 'master' into master

Samuel-DEVULDER committed:
Merge branch 'Samuel-DEVULDER-master'

Samuel-DEVULDER commented:
Hmm. I have a question for the "github" experts.. I still show this present:
![image](https://user-images.githubusercontent.com/33961444/71414909-724fd200-2659-11ea-8c3c-a697198a8f98.png)
 whereas I did all the indicated operations (cmdline merge)
![image](https://user-images.githubusercontent.com/33961444/71414952-a5926100-2659-11ea-89e0-076074427a2b.png)

Question: What did I do wrong? Or what shall I do next?

MBeijer commented:
@Samuel-DEVULDER that’s not what you should do. You commit to your branch and push and the changes will show up in this PR. The merge is for when it’s ready to be merged into the AmigaPorts repo. I have blocked merging unless all reviewers say it’s okay, exactly for this reason to avoid mistakes. Only I can override the review process.

Samuel-DEVULDER committed:
removed BANK which didn't work (got garbage on screen when it is on)

MBeijer commented on SourceS/sdl_compat.h:
```diff
@@ -1,112 +1,100 @@
-	return 0;
-#endif
-}
+// Compatibility wrappers for SDL 1 & 2.
```
Can you fix the line endings in this file?

Samuel-DEVULDER replied:
ok changed LF end-of-line to CR-LF.

MBeijer commented:
@AJenbo @Samuel-DEVULDER A lot of the Vampire includes seems generic enough to be maintained as an external library rather than part of the actual devilutionX project. Perhaps we should set up a new repository for these files and add it as a submodule so it's easier to maintain in the future?

Samuel-DEVULDER committed:
fixed cr/lf

Samuel-DEVULDER commented:
Actually the platform/amiga/vampire folder should be part of the compiler, just like SDL and free-type, free-type2. I have placed it there for simplicity (I don't remind where did I find these.)

MBeijer commented:
> Actually the platform/amiga/vampire folder should be part of the compiler, just like SDL and free-type, free-type2. I have placed it there for simplicity (I don't remind where did I find these.)

@Samuel-DEVULDER yeah, and by doing a vampirelib repository and patching libSDL12 repository with fixes, this will be possible to add to the cross tools, and all future ports to Vampire can benefit of it. :) The vampire override for SDL_flip and SDL_blitsurface could be overridden directly in the SDL code imo. There's already Vampire specific code in there.

Samuel-DEVULDER commented:
Overriding SDL_flip is very "diablo"-specific since I observed that SDL (even with Arczi's modifications) still copies lots of data between hidden and visible surface. So I tuned up a Diablo-specific way of properly using SAGA triple-buffering feature. There are still few SDL_Blit being called for the "viewpart" (the main game potion of the screen) which is a waste of cpu-cycles (cots a few fps I think). I have a branch on this locally, not working as I want yet, but promissing. More to come soon.

MBeijer commented on SourceS/sdl_compat.h:
```diff
@@ -96,8 +82,10 @@ inline int SDLC_SetSurfaceAndPaletteColors(SDL_Surface *surface, SDL_Palette *pa
 	#if SDL1_VIDEO_MODE_BPP == 8
-		// When the video surface is 8bit, we need to set the output pallet as well.
-		SDL_SetColors(SDL_GetVideoSurface(), colors, firstcolor, ncolors);
+
```
```suggestion

```

Samuel-DEVULDER committed:
faster vampire_SurfaceFlip()

Samuel-DEVULDER committed:
Update SourceS/sdl_compat.h

Co-Authored-By: Marlon Beijer <marlon@eevul.org>

Samuel-DEVULDER committed:
just in case

Samuel-DEVULDER committed:
ok this one works

Samuel-DEVULDER committed:
fastest ever!

Samuel-DEVULDER committed:
Merge branch 'master' of https://github.com/Samuel-DEVULDER/devilutionX

Samuel-DEVULDER committed:
faster version of bitstream

Samuel-DEVULDER committed:
faster version of bitstream

Samuel-DEVULDER committed:
faster version of bitstream

Samuel-DEVULDER committed:
added comment

Samuel-DEVULDER committed:
fixed video not being vertically centered

Samuel-DEVULDER committed:
made fps round to closest integer + better C code for gcc/68k in renderRow()

Samuel-DEVULDER committed:
fixed new EA macro

Samuel-DEVULDER committed:
Merge branch 'master' of https://github.com/Samuel-DEVULDER/devilutionX into test-byte-index-special-mode

Samuel-DEVULDER committed:
fixed code format

Samuel-DEVULDER committed:
first successfull compil

Samuel-DEVULDER committed:
works under uae, need to test on real machine

Samuel-DEVULDER committed:
simplified screen-bounds checking code

Samuel-DEVULDER committed:
fixed some transparent tiles not being transparent

Samuel-DEVULDER committed:
improved multithreading in smk_bs_read_1

Samuel-DEVULDER committed:
better handling of selection of RenderLineX functions. Simplified setup code.

Samuel-DEVULDER committed:
fixed typo in memcpy/ammx resulting in possible mem-override

Samuel-DEVULDER committed:
improved scheduling

Samuel-DEVULDER committed:
now uses a 5sec basetime to get a more stable & accurate fps

Samuel-DEVULDER committed:
improved end of mask handling

Samuel-DEVULDER committed:
improved end of mask handling

Samuel-DEVULDER committed:
possibly fixed progress without background

Samuel-DEVULDER commented:
> > ```
> > 1. Solution: add a flag to indicated that the mutex is valid or not.
> > ```
> 
> Do you feel this might be worth upstreaming or is is very Amiga specific?

I get back to this question because I have now a more precise answer. Actually I've dicovered that this is an amiga-gcc issue, see this: https://github.com/bebbo/amiga-gcc/issues/146. Once Bebbo will fix this, we can get rid of this "valid" flag.

Also notice that that issue also affect the game exit when there is no diablo.ini file because the cleanup code for sound accesses the "ini" object which has been destroyed before atexit-registered function were executed (c++ should desroy static object afer atexit()-registered function have been executed.)

AJenbo commented:
Thanks, that really helpful to know. P.s. adjusting the music setting will update the ini while the game is running, don't know if that's useful to you.

Samuel-DEVULDER committed:
code reorg for m68k

Samuel-DEVULDER committed:
removed valid flag now gcc has been fixed

Samuel-DEVULDER committed:
lock/unlock earlier only when running saga

Samuel-DEVULDER committed:
// change the palette the closest to screen wap, otherwise since decodign frames
// takes time, we see the palette change on the previous image typically resutling
// in white spots here and there

Samuel-DEVULDER committed:
reintroduced valid-flag

Samuel-DEVULDER committed:
fixed savefile.

Samuel-DEVULDER committed:
Merge branch 'byteorder-fix-tmp'

now save&load seem to work ok.

Samuel-DEVULDER committed:
fixed Abort() when dtartign mutliplayer.

Samuel-DEVULDER committed:
fixed gold at 5000

Samuel-DEVULDER committed:
fixed right side of pillars being displayed at full intensity in dungeons

Samuel-DEVULDER committed:
fixed endianess issue in SetMapMonsters (which created enforcer hits on the amiga)

AJenbo commented:
Can confirm https://github.com/AmigaPorts/devilutionX/pull/46/commits/5974e965d505196e2e4d956e7dfbc7234c42e988 matches up with the Mac port. Will upstream

Samuel-DEVULDER committed:
removed the 'valid' flag

Samuel-DEVULDER committed:
now uses dlmalloc/dlfree on the global level. This might help finding the out-of-mem issue.

AJenbo commented:
I fixed screenshots for big-endian: https://github.com/diasurgical/devilutionX/pull/554
The save game fixes have also been merged upstream, thanks for working on that.

Samuel-DEVULDER committed:
mertged with upstream

Samuel-DEVULDER committed:
merged with upstream

Samuel-DEVULDER committed:
fixed invisible cursor in-game

Samuel-DEVULDER committed:
fixed '-x' cmdline flag (force windowed mode) not working on saga

Samuel-DEVULDER committed:
added assert_failed() to compile NDEBUG mode

Samuel-DEVULDER committed:
speedup of Cl2BlitSafe() with AMMX & 68k asm. Cpu usage of this fcn drop from 7-8% down to 4%

Samuel-DEVULDER committed:
reformat source code a bit

Samuel-DEVULDER committed:
fixed modulo type

Samuel-DEVULDER committed:
fixed modulo type

Samuel-DEVULDER committed:
fixed stupid erro making the exe crahs on non-vamp amigas

Samuel-DEVULDER committed:
removed bank with makes vampire freeze

Samuel-DEVULDER committed:
added ascii-art for aros

Samuel-DEVULDER committed:
added ascii-art for aros

Samuel-DEVULDER committed:
increased fps why using AMMX version of SDL_MixAudio_m68k_S16MSB

Samuel-DEVULDER committed:
software saturation add now works

Samuel-DEVULDER committed:
added ascii-art for aros

Samuel-DEVULDER committed:
made aros use 320240 screen in place of 320x200 wich seem buggy

Samuel-DEVULDER committed:
added ascii-art for aros

Samuel-DEVULDER committed:
faster asm implem of vaddssw. might fix click noise when walking with only bg music

Samuel-DEVULDER committed:
another attempt at removeing stripes on polish version
